Feat/GitHub actions

gastonva opened the pull request:
## Type of change

* [ ] Feature

## Description of the change
Add a Github Actions CI piplinte that:

- Run linters & formatters, make the CI fail if lint fails
- Run tests
- Measure coverage ([coverage](https://coverage.readthedocs.io/en/7.2.7/)), TBD: make the pipeline fail if coverage drops to 80% (we can change that later)
- Build docker image, make the pipeline fail if the build fails


## Related PRs
N/A

pgrill commented on .github/workflows/python-app.yml:
```diff
@@ -0,0 +1,99 @@
+    runs-on: ubuntu-latest
+    needs: tests
+    steps:
+      - name: Checkout code
```
I am not sure if we need to execute the docker buid in the CI steps. We are building a docker image that will not be used in any place. Maybe we can remove this piece of code and include it later when we manage the CD part of the pipeline.

gastonva replied:
I added that step because that was what the ticket asked for ([ticket](https://www.notion.so/xmartlabs/6a5539b68ff64df79d820299e84a3734?v=b77ed84156d040a8aa635146385ee37c&p=16c6aaa74120491cb643d7e082ac0ad8&pm=s)) but I can remove it if you want.

pgrill replied:
Ohh, I see. The idea is check that the build works and fail if not.
Maybe we can include the caching of the docker layers to improve the performance of this step

pgrill commented on .github/workflows/python-app.yml:
```diff
@@ -0,0 +1,99 @@
+
+jobs:
+  lint-and-format:
+    name: Run Linters & Formatters
```
We have another [PR](https://github.com/xmartlabs/python-template/pull/16) that changes black for ruff and includes the usage of `pre-commit`. Maybe we can merge the work of both PRs in the same one.
What do you think?

pgrill replied:
Personally I prefer ruff over black. The supported rules are similar and the velocity of ruff is much higher than black.

gastonva replied:
Yes, my mind just messed me up, I meant ruff and I agree with you. Sorry. I'll add ruff and `pre-commit` in the next commit.

gastonva commented:
I merged everything from #16. I made some changes to get the `mypy` hook working and added all the necessary settings to the `pyproject.toml`, following the official documentation. I also added a cache layer to the `docker-build` step in the GitHub Actions CI workflow.

cc: @pgrill

m-revetria commented on .coverage:
this should be ignored from git right?

m-revetria commented on .github/workflows/python-app.yml:
```diff
@@ -0,0 +1,114 @@
+
+      - name: Run tests with coverage
+        run: |
+          poetry run coverage run -m pytest
```
doesn't this need running migration before?

gastonva replied:
No, because `src/tests/base.py` is using `SQLBase.metadata.create_all(bind=engine)` to create the tables and then some tests creates some default entries.

m-revetria commented on .github/workflows/python-app.yml:
```diff
@@ -0,0 +1,114 @@
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-docker-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-docker-
```
is something missing here -at the end of `-docker-`? I'm asking cuz the leading `-` looks suspicious

m-revetria commented on .github/workflows/python-app.yml:
```diff
@@ -0,0 +1,114 @@
+
+      - name: Run linters and formatters
+        run: |
+          poetry run python -m ruff check --fix src
+          poetry run python -m ruff format --check src
+          poetry run python -m mypy src --exclude 'src/alembic/'
```
is pre commit is already installed, can't we use it here?

gastonva replied:
Great question, I think we can, let me try.

m-revetria commented on pyproject.toml:
```diff
@@ -5,19 +5,20 @@ version = "0.1.0"
 alembic = "^1.15.1"
 asyncpg = "^0.30.0"
 bcrypt = "4.3.0"
+coverage = "^7.7.1"
```
this should go to the `dev` group right?
